python/lsst/ts/atpneumaticssimulator/pneumatics_simulator.py

```diff
@@ -239,6 +387,179 @@ async def write_noack_response(self, sequence_id: int) -> None:
+            cover1ClosedActive=self.m1_cover_limit_switches.cover1ClosedActive,
+            cover2ClosedActive=self.m1_cover_limit_switches.cover2ClosedActive,
+            cover3ClosedActive=self.m1_cover_limit_switches.cover3ClosedActive,
+            cover4ClosedActive=self.m1_cover_limit_switches.cover4ClosedActive,
+            cover1OpenedActive=self.m1_cover_limit_switches.cover1OpenedActive,
+            cover2OpenedActive=self.m1_cover_limit_switches.cover2OpenedActive,
+            cover3OpenedActive=self.m1_cover_limit_switches.cover3OpenedActive,
+            cover4OpenedActive=self.m1_cover_limit_switches.cover4OpenedActive,
+        )
+        if opened and closed:
+            self.m1_cover_state = ATPneumatics.MirrorCoverState.INVALID
+        elif opened:
+            self.m1_cover_state = ATPneumatics.MirrorCoverState.OPENED
+        elif closed:
+            self.m1_cover_state = ATPneumatics.MirrorCoverState.CLOSED
+        else:
+            self.m1_cover_state = ATPneumatics.MirrorCoverState.INMOTION
+        await self._write_evt(evt_id=Event.M1COVERSTATE, state=self.m1_cover_state)
+
+    async def telemetry_loop(self) -> None:
+        """Output telemetry and events that have changed
+
+        Notes
+        -----
+        Here are the telemetry topics that are output:
+
+        * m1AirPressure
+        * m2AirPressure
+        * mainAirSourcePressure
+        * loadCell
+
+        See `update_events` for the events that are output.
+        """
+        while True:
+            await self.update_telemetry()
+            await asyncio.sleep(self._telemetry_interval)
+
+    async def update_telemetry(self) -> None:
+        """Output all telemetry data messages."""
+        try:
+            opened_state = ATPneumatics.AirValveState.OPENED
+            main_valve_open = self.main_valve_state == opened_state
+
+            if main_valve_open and self.m1_state == opened_state:
+                self.m1_air_pressure.pressure = self.m1_pressure
+            else:
+                self.m1_air_pressure.pressure = 0
+
+            if main_valve_open and self.m2_state == opened_state:
+                self.m2_air_pressure.pressure = self.m2_pressure
+            else:
+                self.m2_air_pressure.pressure = 0
+
+            await self._write_telemetry(
+                tel_id=Telemetry.M1_AIR_PRESSURE,
+                data=self.m1_air_pressure,
+            )
+            await self._write_telemetry(
+                tel_id=Telemetry.M2_AIR_PRESSURE,
+                data=self.m2_air_pressure,
+            )
+            await self._write_telemetry(
+                tel_id=Telemetry.MAIN_AIR_SOURCE_PRESSURE,
+                data=self.main_air_source_pressure,
+            )
+            await self._write_telemetry(
+                tel_id=Telemetry.LOAD_CELL,
+                data=self.load_cell,
+            )
+        except Exception as e:
+            print(f"update_telemetry failed: {e}")
+            raise
+
```

Comment:
It's technical debt (my old code), but please consider deleting the `__enter__` method (the next few lines).